File: ruby/array_test.rb
require 'minitest/autorun'

describe "array" do

  it "create array" do
    assert_equal ["apple", "orange"], ["apple", "orange"].freeze
  end

  it "create array mutable" do
    assert_equal ["apple", "orange"], %w{apple orange}
    assert_equal ["apple", "orange"], "apple orange".split
    assert_equal ["apple", "orange"], ["apple", "orange"]
  end

  it "add mutate" do
    array = [1, 2]
    array.insert(1, 10)
    array << 3
    assert_equal [1, 10, 2, 3], array
  end

  it "get object" do
    assert_equal "apple", ["apple", "orange"][0]
  end

  it "get index" do
    assert_equal 0, ["apple", "orange"].index("apple")
    assert_equal nil, ["apple", "orange"].index("not found")
  end

  it "include" do
    assert_equal true, ["apple", "orange"].include?("apple")
  end

  it "remove mutate" do
    array = [1, 2, 3, 4, 4]
    array.delete_at(0)
    array.delete(4)
    assert_equal [2, 3], array
  end

  it "each" do
    sum = 0
    [1, 2].each {|x| sum += x}
    assert_equal 3, sum
  end

  it "each with index" do
    sum = 0
    [1, 2].each_with_index {|x, i| sum += x * i}
    assert_equal 2, sum
  end

  it "map" do
    assert_equal [2, 3], [1, 2].map {|x| x + 1}
  end

  it "map with index" do
    assert_equal [1, 3], [1, 2].each_with_index.map {|x, i| x + i}
  end

  it "inject" do
    assert_equal 3, [1, 2].inject(0) {|sum, x| sum + x}
  end

  it "inject with index" do
    assert_equal 2, [1, 2].each_with_index.inject(0) {|sum, (x, i)| sum + x * i}
  end

  it "select" do
    assert_equal [2], [1, 2].select {|x| x > 1}
  end

  it "select mutate" do
    array = [1, 2]
    array.select! {|x| x > 1}
    assert_equal [2], array
  end

  it "compact" do
    assert_equal [1, 2], [1, nil, 2].compact
  end

  it "compact mutate" do
    array = [1, nil, 2]
    object_id = array.object_id
    array.compact!
    assert_equal [1, 2], array
    assert_equal object_id, array.object_id
  end

  it "sort" do
    assert_equal [1, 2], [2, 1].sort
    assert_equal [2, 1], [1, 2].sort_by {|x| -x}
    assert_equal [[1, 3], [1, 2], [2, 1]],
      [[2, 1], [1, 3], [1, 2]].sort_by {|x| [x.first, -x.last]}
    assert_equal [1, 2], [2, 1].sort {|a, b| b == a ? 0 : (a > b ? 1 : -1)}
  end

end

File: elixir/test/array_test.exs
ExUnit.start

defmodule ArrayTest do
  use ExUnit.Case, async: true

  test "create array" do
    assert ["apple", "orange"] = ["apple", "orange"]
  end

  test "add" do
    # http://elixir-lang.org/docs/v1.0/elixir/List.html
    array = List.insert_at([1, 2], 1, 10)
    array = array ++ [3]
    assert [1, 10, 2, 3] = array
  end

  test "get object" do
    # http://elixir-lang.org/docs/v1.0/elixir/Enum.html
    assert "apple" = Enum.at(["apple", "orange"], 0)
  end

  test "get index" do
    assert 0 = Enum.find_index(["apple", "orange"], fn(x)-> x == "apple" end)
    assert nil == Enum.find_index(["apple", "orange"], &(&1 == "not found"))
  end

  test "include" do
    assert Enum.member?(["apple", "orange"], "apple")
  end

  test "remove" do
    array = [1, 2, 3, 4, 4]
    |> List.delete_at(0)
    |> Enum.reject(&(&1 == 4))
    assert [2, 3] = array
  end

  test "each" do
    # Not useful in elixir
    # Enum.each([1, 2], &(IO.puts(&1)))
  end

  test "each with index" do
    # Not useful in elixir
    # array = [1, 2]
    # |> Enum.with_index
    # |> Enum.each(fn {x, i} -> IO.puts "#{x} at #{i}" end)
  end

  test "map" do
    assert [2, 3] = Enum.map([1, 2], &(&1 + 1))
  end

  test "map with index" do
    array = [1, 2]
    |> Enum.with_index
    |> Enum.map(fn {x, i} -> x + i end)
    assert [1, 3] = array
  end

  test "inject" do
    assert 3 = Enum.reduce([1, 2], 0, fn(x, sum) -> x + sum end)
  end

  test "inject with index" do
    assert 2 = [1, 2]
    |> Enum.with_index
    |> Enum.reduce(0, fn ({x, i}, sum) -> sum + x * i end)
  end

  test "select" do
    assert [2] = Enum.filter([1, 2], &(&1 > 1))
  end

  test "compact" do
    assert [1, 2] = Enum.reject([1, nil, 2], &(&1 == nil))
  end

  test "sort" do
    assert [1, 2] = Enum.sort([2, 1])
    assert [2, 1] = Enum.sort_by([1, 2], &(-&1))
    assert [[1, 3], [1, 2], [2, 1]] =
      Enum.sort_by([[2, 1], [1, 3], [1, 2]], fn([a, b]) -> [a, -b] end)
  end

end
